Applications/Spire/Source/Ui/DropDownBox.cpp

```diff
@@ -0,0 +1,194 @@
+#include "Spire/Ui/DropDownBox.hpp"
+#include <QCoreApplication>
+#include <QHBoxLayout>
+#include <QVBoxLayout>
+#include <QKeyEvent>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ArrayListModel.hpp"
+#include "Spire/Ui/Button.hpp"
+#include "Spire/Ui/CustomQtVariants.hpp"
+#include "Spire/Ui/DropDownList.hpp"
+#include "Spire/Ui/Icon.hpp"
+#include "Spire/Ui/ListItem.hpp"
+#include "Spire/Ui/ListView.hpp"
+#include "Spire/Ui/TextBox.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  auto DEFAULT_STYLE() {
+    auto style = StyleSheet();
+    style.get(Any() >> is_a<Icon>()).
+      set(BackgroundColor(QColor::fromRgb(0, 0, 0, 0))).
+      set(Fill(QColor::fromRgb(0x33, 0x33, 0x33)));
+    style.get(Disabled() >> is_a<Icon>()).
+      set(Fill(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    style.get(ReadOnly() >> is_a<Icon>()).
+      set(Visibility::NONE);
+    style.get(Any() >> (is_a<TextBox>() && !(+Any() << is_a<ListView>()))).
+      set(PaddingRight(scale_width(8)));
+    style.get(Disabled() >> (is_a<TextBox>() && !(+Any() << is_a<ListView>()))).
+      set(TextColor(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    return style;
+  }
+}
+
+DropDownBox::DropDownBox(ListView& list_view, QWidget* parent)
+    : QWidget(parent),
+      m_list_view(&list_view),
+      m_is_read_only(false) {
+  auto container = new QWidget();
+  auto container_layout = new QHBoxLayout(container);
+  container_layout->setContentsMargins({});
+  container_layout->setSpacing(0);
+  m_text_box = make_label("");
+  m_text_box->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  m_text_box->setAttribute(Qt::WA_TransparentForMouseEvents);
+  container_layout->addWidget(m_text_box);
+  auto icon_layout = new QVBoxLayout();
+  icon_layout->setContentsMargins({});
+  icon_layout->setSpacing(0);
+  icon_layout->addStretch();
+  auto drop_down_icon = new Icon(
+    imageFromSvg(":/Icons/dropdown-arrow.svg", scale(6, 4)));
+  drop_down_icon->setFixedSize(scale(6, 4));
+  drop_down_icon->setFocusPolicy(Qt::NoFocus);
+  icon_layout->addWidget(drop_down_icon);
+  icon_layout->addStretch();
+  container_layout->addLayout(icon_layout);
+  m_input_box = make_input_box(container);
+  m_button = new Button(m_input_box);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(m_button);
+  proxy_style(*this, *m_input_box);
+  set_style(*this, DEFAULT_STYLE());
+  setFocusProxy(m_button);
+  m_drop_down_list = new DropDownList(*m_list_view, this);
+  auto list_model = m_list_view->get_list_model();
+  for(auto i = 0; i < list_model->get_size(); ++i) {
+    if(auto item = m_list_view->get_list_item(i)) {
```

Comment:
Is it necessary to check that the item at index _n_ isn't null if the model contains at least _n_ items? Is there a case where a value in the model won't have an item in the view?